### PR TITLE
Credit Card Escrow

### PR DESCRIPTION
**Pull request by end-r**

### CODE REVIEW

File: contracts/platform/contracts/escrow/ERC20Escrow.sol [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    }
+
+    /**
+     * @dev Create a new escrow vault
+     *
+     * @param vault the escrow vault to be created
+     * @param from the address from which to pull the tokens
+     */
+    function escrow(Vault memory vault, address from) public returns (uint256 vaultID) {
+        require(!mutexLocked, "mutex must be unlocked");
```
Comment (kermankohli) [inline contracts/platform/contracts/escrow/ERC20Escrow.sol]: Same above regarding the require statements. Putting the name of the contract is massively helpful when debugging.

File: contracts/platform/contracts/escrow/ERC721Escrow.sol [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+     * @dev Set whether a particular contract has batch transfers available
+     *
+     * @param asset the address of the asset contract
+     * @param enabled whether this asset can use batch transfers
+     */
+    function setSingleTxEnabled(address asset, bool enabled) external onlyOwner {
+        singleTxEnabled[asset] = enabled;
+    }
+
+    function _transfer(uint256 vaultID, address from, address to) internal;
```
Comment (kermankohli) [inline contracts/platform/contracts/escrow/ERC721Escrow.sol]: Is this meant to be an abstract contract? If so, can we rename it to make it clear that it is? A simple suggestion alternatively is to call something externally and make it adhere to an interface.

File: contracts/platform/contracts/escrow/ERC721Escrow.sol [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        require(!_areAnyAssetsEscrowed(vaultID), "must own none of the tokens");
+        require(!mutexLocked, "mutex must be unlocked");
+        mutexLocked = true;
+        // solium-disable-next-line security/no-low-level-calls
+        callbackTo.call(callbackData);
+        require(_areAllAssetsEscrowed(vaultID), "must now own all tokens");
+        mutexLocked = false;
+    }
+
+    function _existsAndOwnedBy(address asset, uint256 tokenID, address possibleOwner) internal returns (bool) {
```
Comment (kermankohli) [inline contracts/platform/contracts/escrow/ERC721Escrow.sol]: Formatting of function please :)

File: contracts/platform/__tests__/escrow/erc20.test.ts [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+import { BigNumberish } from 'ethers/utils';
+
+const provider = new ethers.providers.JsonRpcProvider();
+const blockchain = new Blockchain();
+
+const ZERO_EX = '0x0000000000000000000000000000000000000000';
+
+describe('ERC20Escrow', () => {
+
+  const [user, other] = generatedWallets(provider);
```
Comment (kermankohli) [inline contracts/platform/__tests__/escrow/erc20.test.ts]: Typically you should create all the users you want to test with:

ownerWallet, userWallet, unauthorisedWallet, immutableWallet

Then have test cases which show interactions with all of these parties

File: contracts/platform/__tests__/escrow/list-erc721.test.ts [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+    it('should be able to deploy the escrow contract', async () => {
+      const escrow = await new ListERC721EscrowFactory(user).deploy();
+    });
+  });
+
+  describe('#escrow', () => {
+
+    let escrow: ListERC721Escrow;
+    let erc721: TestERC721Token;
+
```
Comment (kermankohli) [inline contracts/platform/__tests__/escrow/list-erc721.test.ts]: Subject pattern here

File: contracts/platform/contracts/escrow/ERC721Escrow.sol [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+        bytes memory data = abi.encodeWithSignature("ownerOf(uint256)", tokenID);
+        // solium-disable-next-line security/no-low-level-calls
+        (bool success, bytes memory response) = asset.call(data);
+        if (success) {
+            address owner;
+            // solium-disable-next-line security/no-inline-assembly
+            assembly {
+                owner := mload(add(response, 20))
```
Comment (kermankohli) [inline contracts/platform/contracts/escrow/ERC721Escrow.sol]: Any particular reason for using assembly here?

File: contracts/platform/contracts/escrow/ListERC721Escrow.sol [hunk 7]
```diff
@@ -0,0 +1,90 @@
+pragma solidity 0.5.11;
+pragma experimental ABIEncoderV2;
+
+import "./ERC721Escrow.sol";
+import "../token/ListTransfer.sol";
+import "@openzeppelin/contracts/token/ERC721/IERC721.sol";
+import "./IListERC721Escrow.sol";
+
+contract ListERC721Escrow is IListERC721Escrow, ERC721Escrow {
```
Comment (kermankohli) [inline contracts/platform/contracts/escrow/ListERC721Escrow.sol]: Something about the naming of this inheritance structure leaves me very confused as to what this contract is inheriting from and what it's meant to do. Is there a use case for using an interface for IListERC721Escrow? If not then we should limit the use case of our interfaces unless there's an explicit need.

File: contracts/platform/contracts/escrow/IERC20Escrow.sol [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+     * @param vault the escrow vault to be created
+     * @param callbackTo the address to use for the callback transaction
+     * @param callbackData the data to pass to the callback transaction
+     */
+    function callbackEscrow(
+        Vault memory vault, address callbackTo, bytes memory callbackData
+    ) public returns (uint256 vaultID);
+
+    /**
+     * @dev Create a new escrow vault
```
Comment (kermankohli) [inline contracts/platform/contracts/escrow/IERC20Escrow.sol]: The description of this is the same as the above function's description

File: contracts/platform/contracts/escrow/BatchERC721Escrow.sol [hunk 9]
```diff
@@ -0,0 +1,90 @@
+pragma solidity 0.5.11;
+pragma experimental ABIEncoderV2;
+
+import "../token/BatchTransfer.sol";
+import "./ERC721Escrow.sol";
+import "@openzeppelin/contracts/token/ERC721/IERC721.sol";
+import "./IBatchERC721Escrow.sol";
+
+contract BatchERC721Escrow is IBatchERC721Escrow, ERC721Escrow {
```
Comment (kermankohli) [inline contracts/platform/contracts/escrow/BatchERC721Escrow.sol]: Formatting of require statements, also see comment below regarding use of interfaces and simplifying this inheritance structure.

File: contracts/platform/contracts/escrow/ListERC721Escrow.sol [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+    Vault[] public vaults;
+
+    /**
+     * @dev Create an escrow account where assets will be pushed into escrow by another contract
+     *
+     * @param vault the details of the new escrow vault
+     * @param callbackTo the address to use for the callback transaction
+     * @param callbackData the data to pass to the callback transaction
+     */
+    function callbackEscrow(Vault memory vault, address callbackTo, bytes memory callbackData) public returns (uint256 vaultID) {
```
Comment (kermankohli) [inline contracts/platform/contracts/escrow/ListERC721Escrow.sol]: Might be worth to run the solidity linter to make sure things like this are caught out

File: contracts/platform/contracts/escrow/releaser/CreditCardEscrow.sol [hunk 11]
```diff
@@ -0,0 +1,286 @@
...
+        delete locks[_id];
+
+        emit Destroyed(_id);
+    }
+
+    /**
+     * @dev Escrow some amount of ERC20 tokens
+     *
+     * @param _vault the details of the escrow vault
+     * @param _cbTo the address to use for the callback transaction
```
Comment (kermankohli) [inline contracts/platform/contracts/escrow/releaser/CreditCardEscrow.sol:248]: Proper names please `callbackTo`

**Comment by kermankohli** [comment]
Feel free to merge in once that small thing is changed :)

File: contracts/platform/contracts/escrow/releaser/CreditCardEscrow.sol (hunk 11 above)
Comment (end-r) [reply contracts/platform/contracts/escrow/releaser/CreditCardEscrow.sol]: Resolved!

**Pull request merged**